### DISCUSSION TITLE
How granular the libs should be?

### DISCUSSION THREAD

**Discussion by J-Huang** [nrwl/nx#8955] [General]
In [the documentation](https://nx.dev/structure/creating-libraries#1-faster-commands), it says:

> The more granular your libraries are, the more effective nx affected and Nx's computation cache will be.

It makes perfect sense since lib is the smallest unit in nx world.
Does that mean it recommends to separate tests from the lib as another lib? I would imagine most users, including us, colocate tests with the real code in the same lib. If a test imports a module, which has not been imported by the lib itself, it would trigger unnecessary tests. The following demo graph shows the idea.
![Screen Shot 2022-03-05 at 7 29 01 PM](https://user-images.githubusercontent.com/4263459/156908022-8c54ecd6-a9bd-462d-8277-f18ea7cb1863.png)

If a library is instead considered a lib and a lib-tests (on the right side of the demo graph), change graph behavior is clearly desirable. Since this could be an issue that many users encounter, wondering if there is already a recommended pattern.
- Are there formal patterns that exist?
- Does NX apply such a pattern, even if informally? Why/why-not?
- A broader question is, should lib be as slim as possible, even if that means there will be thousands of libs?

cc: @vsavkin  @cdaringe

**Comment by cdaringe** [comment]
I suppose I drew the arrow backwards from `Lib B -> Lib B __tests__` in the right diagram :)
